### PR TITLE
Add ability to support attrs in Panel and by extension FieldPanel #10133 

### LINKED ISSUE #10133: Add ability to support `attrs` in `Panel` and by extension `FieldPanel`
### Is your proposal related to a problem?

With the adoption of Stimulus under way and plans for developer facing documentation (non-contributors) coming up soon, it would be good to be able to provide a concise way to allow for HTML attributes to be easily passed from `Panel`/`FieldPanel` through to the HTML template.

This would allow for official support for much simpler customisations where using/adding data attributes is needed (e.g. `data-controller="custom-controller"`) or even common requirements like aria or ID attributes.

### Describe the solution you'd like

* `Panel` class should allow a new kwarg `attrs` - https://github.com/wagtail/wagtail/blob/main/wagtail/admin/panels/base.py#L48 (including support for cloning this in `clone_kwargs`)
* `FieldPanel` should support have something similar to `BASE_ATTRS` on the class as a property, similar to `TEMPLATE_VAR` https://github.com/wagtail/wagtail/blob/main/wagtail/admin/panels/field_panel.py#L12
* `FieldPanel` could have `BASE_ATTRS` set to `{"data-field-wrapper": true}` by default and then on the `FieldPanel` init, it sets this as the `attrs`, and extends it with any `attrs` provided to the `FieldPanel` instance, before calling `super`. Note that someone should be able to override the `data-field-wrapper` if they chose (although this would probably break things).
* `BoundPanel.get_context_data` should support passing in the `attrs` to the template (probably at the root `Panel` implementation.
* The template `wagtail/admin/templates/wagtailadmin/shared/field.html` should support rendering the attributes in the same way that Django attrs are rendered.
* Extending `FieldPanel` or `Panel` and the BoundPanel's `get_context` should easily allow appending/adjusting the `attrs` through the inheritance chain as needed.

### Example usage

Trivial example but shows off how it would work

```py
# models.py
class OperatingHours(models.Model):

    day = models.CharField(max_length=4, choices=DAY_CHOICES, default="MON")

    panels = [
        FieldPanel("day", attrs={"data-controller": "today"}),
    ]
```

```js
// .../static/js/TodayController.js - added via hooks

const FilterController = window.Stimulus.constructor.createController({
  connect() {
    const field = this.element.querySelector('[data-field] input');
    const todayWeekday = 3; // get today's weekday as a number
    if (field.value === todayWeekday) {
        this.element.classList.add('highlight');
    }
  }
});

window.Stimulus.register('today', FilterController);
```

#### `attrs` rendering should emulate Django widget attrs

Using Django's terminology and ways of working with widget attrs on panels feels intuitive and is also very powerful, note that this is not a proposal to allow for non-called `Panel/FieldPanel` as is done with Widgets.

The way that `attrs` in Django widgets get rendered is a bit nuanced and has some really nice defaults, the main one being how True/False items are handled. True will just add the attribute as is (no value) and False will not add the attribute at all. There are also the built in escaping benefits that come with this being driven at the Django side.

There is also some nice handling of cloning/extending the `attrs` object.

* https://docs.djangoproject.com/en/4.1/ref/forms/widgets/#django.forms.Widget.attrs
* See https://github.com/django/django/blob/main/django/forms/widgets.py#L281
* https://github.com/django/django/blob/main/django/forms/templates/django/forms/attrs.html

### Describe alternatives you've considered

* Wagtail provides a documented way to add attributes to a field, via the `widget=` param on `FieldPanel`, this is often enough to get an attribute inside the element.
* Maybe we do not want to add another kwarg to `Panel`, we already do support `classname` and that may be enough at the root level.
* We may not want to provide the ability to do this at the panel level and instead want to push this kind of thing to the template level, however, our `field_panel` does not allow for appending `attrs` by extension right now so it's very difficult to do this without just adding another `div` around it. Additionally, when putting attrs in the template's HTML, there are further escaping considerations that have to be done manually and add developer overhead / security risks.

### Additional context

* See answer to a problem in a comment - https://github.com/wagtail/wagtail/issues/7417#issuecomment-1438396850 - this would be easier to implement (when that solution ends up in the docs for example).
* `PageChooserPanel`, `InlinePanel` and other custom Panels could much more easily support Stimulus controller attributes without needing to change much/any of the template HTML files if needed.
* We can provide some nice documented / answered solutions to #161 & #1063 via essentially adding some JavaScript snippets and adding `attrs` to the `FieldPanel`.

### PR DESCRIPTION
**Pull request by salty-ivy**
Fixes #10133 

Adds ability to support passing attributes via FieldPanel   ( targeted for stimulus controller attributes )

### How would it work

```python
class HomePage(Page):
    body = RichTextField(blank=True)

    content_panels = Page.content_panels + [
        FieldPanel('body', attrs={"data-controller": "today"}),
    ]
```

### result 
Adds `data-controller="today"` at the root div of the field.
image

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?

### CODE REVIEW

File: wagtail/admin/panels/base.py
Comment (lb-) [inline wagtail/admin/panels/base.py:250]:
```suggestion
```

Let's remove this here, it seems like this is not the approach taken for other similar things in BoundPanel like icon, instead let's create a property like this...

```python
        @property
        def attrs(self):
            return self.panel.attrs
```

Maybe put it above or below the `def icon` on line ~235-ish. Let's try to keep the addition consistent, either we add attrs as the last thing or add it consistently with some context like before/after icon.

File: wagtail/admin/panels/field_panel.py
Comment (lb-) [inline wagtail/admin/panels/field_panel.py:21]: If we use the approach already set in `Panel`, we should only need one single line added here - the 
```python
BASE_ATTRS = { "data-field-wrapper":True }
```
on line 14, the pulling out of `attrs` will be handled without any additional code by the root class implementation.

File: wagtail/admin/templates/wagtailadmin/shared/field.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/field.html:60]: Firstly - I think the requirements are to put these `attrs` on the root div - which would be line 17 in this file, a reminder we are replacing the `data-field-wrapper` attribute with the BASE_ATTRS approach, so where that attribute is is where our new code should go.

```suggestion
<div class="w-field__wrapper {{ classname }}" {% include "wagtailadmin/tables/attrs.html" with attrs=self.attrs %}>
```

Let's use this template `wagtail/admin/templates/wagtailadmin/tables/attrs.html`, but we probably want to move this template to somewhere shared.

Here is a usage example `wagtail/admin/templates/wagtailadmin/tables/title_cell.html`

With the template moved to the shared folder (`wagtail/admin/templates/wagtailadmin/shared/attrs.html`, it should look something like this...

```suggestion
<div class="w-field__wrapper {{ classname }}" {% include "wagtailadmin/shared/attrs.html" with attrs=self.attrs %}>
```

Important: I got this working but it has to use the `self.attrs` not just `attrs` - we may need some help from Matt or others to see if we can improve this but it should be good enough for you to continue your dev and start on tests.

**Review by lb- — COMMENTED** [review]
Great start @salty-ivy  - here is some feedback to hopefully help unblock you.

**Comment by salty-ivy** [comment]
> Great start @salty-ivy - here is some feedback to hopefully help unblock you.

thanks a lot, I guess I should have gone over the links of the additional context, that would have save your time :(

File: wagtail/admin/panels/base.py
Comment (laymonage) [reply wagtail/admin/panels/base.py]: Note that `dict.update()` mutates the dictionary and returns `None`, and class attributes are shared across instances. Also, avoid having mutable default arguments for the same reason.

You probably want to do something like `self.attrs = self.BASE_ATTRS.copy()` and then `if attrs: self.attrs.update(attrs)` instead.

Comment (lb-) [reply wagtail/admin/panels/base.py]: Great call out Sage. Thank you

**Comment by lb-** [comment]
Thanks Aman. I might not get to it for a week or so as I'm away. Others on the core team may get a chance to look in the meantime though.

**Comment by lb-** [comment]
Only quick thought - one more test for adding attrs to Panel would be good, not just Field panel.

**Comment by lb-** [comment]
We should also document this here. https://docs.wagtail.org/en/stable/reference/pages/panels.html#field-panel (FieldPanel)

Additionally this attrs should work for FieldRowPanel, HelpPanel and others. I don't think it will for MultiFieldPanel as that does not have a root element. But, for simplicity see if it's easy to add those in this PR of we can split to a new issue later. 

The main priority is Panel and FieldPanel.

**Comment by salty-ivy** [comment]
> We should also document this here. https://docs.wagtail.org/en/stable/reference/pages/panels.html#field-panel (FieldPanel)
> 
> Additionally this attrs should work for FieldRowPanel, HelpPanel and others. I don't think it will for MultiFieldPanel as that does not have a root element. But, for simplicity see if it's easy to add those in this PR of we can split to a new issue later.
> 
> The main priority is Panel and FieldPanel.

@lb- I think `FieldRowPanel` and `MultiFieldPanel` have almost same implementation. Both of their templates uses `multi_field_panel_child.html`  if render the `attrs`   here https://github.com/wagtail/wagtail/blob/b946b3d66d67e8cd7179c188078c6ed735c46b9c/wagtail/admin/templates/wagtailadmin/panels/multi_field_panel_child.html#L2

wouldn't this work for both of them?

**Comment by salty-ivy** [comment]
For `InlinePanel` I am rendering the `attrs` here https://github.com/wagtail/wagtail/blob/b946b3d66d67e8cd7179c188078c6ed735c46b9c/wagtail/admin/templates/wagtailadmin/panels/inline_panel_child.html#L5

including the `data-contentpath-disabled`  by overriding  `BASE_ATTRS` in `InlinePanel` definitions.

let me know if these are right places.

**Comment by lb-** [comment]
Both those suggested changes look good. Great find.

**Comment by salty-ivy** [comment]
@lb- could you have a look at this one before I move towards the DOCs

**Review by lb- — CHANGES_REQUESTED** [review]
@salty-ivy thank you for this! Most of the functionality is solid, just a bit more polish and a few more template/docs additions before it's ready to merge.

Sorry for the run-around but I think the `InlinePanel` should be solved as a separate PR, it needs a bit of a rethink in the way that the HTML is structured and children are rendered. However, once the core `attrs` usage is in and documented, the feature can be built on for further usage throughout Wagtail admin and I do not want to block this PR by trying to cover every single usage of data attributes on all panels :) 

I have listed the items I have found by reviewing the code and also reviewing `Panel` usage - see below. Hope this helps you polish this off. Note that `read_only` has now come into the `FieldPanel` definition, but do not worry about rebasing if you are not ready to do that, I can rebase once you are ready for the next review.

Ping me when this is good to look at again.

### 0. `self.attrs` vs `attrs`

* I think we need to revisit how the usage of `self.` is being used and ensure that it's passed directly to the template's context.
* Please have a dig into why this was not working initially
* We want to avoid needing to use `self` if we can.


### 1. Comment Panel

* We can add support for CommentPanel in `attrs` and use it for an existing data attribute
* `wagtail/admin/panels/comment_panel.py` -> add `BASE_ATTRS = {"data-annotation": True}`
* `wagtail/admin/templates/wagtailadmin/panels/comments/comment_panel.html` add `<button type="button" id="comment-icon" class="w-field__comment-button w-field__comment-button--reveal u-hidden" {% include "wagtailadmin/shared/attrs.html" with attrs=self.attrs %}>`

### 2. Help Panel

* We can ensure that HelpPanel supports the param by adding one line to the html template
* `wagtail/admin/templates/wagtailadmin/panels/help_panel.html` add `<fieldset {% include "wagtailadmin/shared/attrs.html" with attrs=self.attrs %}>`
* Please include a test for this support

### 3. `TabbedInterface`

* This extends `Panel` and can also use this approach
* `wagtail/admin/panels/group.py` -> `BASE_ATTRS = {"data-tabs": True}`
* `wagtail/admin/templates/wagtailadmin/panels/tabbed_interface.html` -> `<div class="w-tabs" {% include "wagtailadmin/shared/attrs.html" with attrs=self.attrs %}>`

### 4. InlinePanel

* For now, remove the changes to inline panel and the inline panel child template, I think this approach needs some more time and thought to be polished.
* Unfortunately InlinePanel does not have an obvious wrapper HTML element, so we will just need to skip support in that panel in this PR.

### 5. Object list

* Adding `ObjectList` support just means we are closer to supporting all logical template usage of `attrs`
* `wagtail/admin/templates/wagtailadmin/panels/object_list.html` -> `<div class="w-form-width" {% include "wagtailadmin/shared/attrs.html" with attrs=self.attrs %}>`

### 6. Documentation

Once the items above are added and potentially some more unit tests, we do need to ensure the documentation is done, here is a quick snippet example.

Added to `docs/reference/pages/panels.md`, just need to update `FieldPanel` & `HelpPanel`. The other panels will not support the `attrs` simply so no need to list the nuance (i.e. they would need to consider parent/child items in panel groups). This will cover the most common usage.

```md
(field_panel)=
### FieldPanel

```{eval-rst}
.. autoclass:: FieldPanel
    # ... all the others - add after `.. attribute:: FieldPanel.read_only (optional)`

    .. attribute:: FieldPanel.attrs (optional)

        Allows A dictionary containing HTML attributes to be set on the rendered panel. If you assign a value of True or False to an attribute, it will be rendered as an HTML5 boolean attribute.

        It behaves the same way as :doc:`Django's Widget attrs <django:ref/forms/widgets/#django.forms.Widget.attrs>` except that it will output the attributes on the field's container element and not the widget element.

# ...

.. autoclass:: HelpPanel

    .. attribute:: HelpPanel.content

        HTML string that gets displayed in the panel.

    .. attribute:: HelpPanel.template

        Path to a template rendering the full panel HTML.

    .. attribute:: FieldPanel.attrs (optional)

        See `FieldPanel.attrs`. Allows A dictionary containing HTML attributes to be set on the rendered panel.
```

### 7. Add pydoc string to `Panel`

`wagtail/admin/panels/base.py`

```py
class Panel:
    """
    Defines part (or all)....
    :param attrs: A dictionary of HTML attributes to add to the panel's HTML element.
    """
```


### 8. Note about others

Both of the following panel templates do not render any markup/html so no need to consider `attrs` usage in those.  Just adding a reminder for reference.

* `FieldRowPanel` -> `wagtailadmin/panels/field_row_panel.html`
* `MultiFieldPanel` -> `wagtail/admin/templates/wagtailadmin/panels/multi_field_panel.html`

**Comment by salty-ivy** [comment]
@lb-  Thanks for the comment and review, I have already re-based this branch with `read-only` functionality quite a while ago.
so I would be working on your suggestions directly.

**Comment by lb-** [comment]
Thanks @salty-ivy

**Comment by lb-** [comment]
@salty-ivy as per Slack, you're still keen on pursuing this but may not get back to it straight away.

**Comment by pySilver** [comment]
Oh I need this functionality so much! I'm glad to see it almost there. 

Quick note: we need an ability to set attributes for whole page edit form (or other form in case we are using ModelAdmin or Snippets). This is very useful for cases where stimulus controller should be applied on a higher level (node that wraps all form fields)

File: wagtail/admin/panels/base.py
Comment (pySilver) [inline wagtail/admin/panels/base.py:1]: There is an outdated import of `text_from_html`.  It should be imported from `from wagtail.admin.compare import text_from_html`

**Comment by pySilver** [comment]
@salty-ivy Just asking if you are willing to complete the PR. If you need any help here I'll be happy to help.

**Comment by salty-ivy** [comment]
@pySilver Thanks a lot for asking and real big apologies for all this delay. I would be getting back to this PR
asap, as I got selected by wagtail in GSoC project https://github.com/orgs/wagtail/projects/30 and examsss.

I would be free by 26th and can start working back on it. We can work this out.

if you don't wanna wait till 26th,  you are very much welcome to take over me : )

**Comment by salty-ivy** [comment]
Great !

**Comment by lb-** [comment]
Thanks both for the discourse on this. 

> Quick note: we need an ability to set attributes for whole page edit form (or other form in case we are using ModelAdmin or Snippets). This is very useful for cases where stimulus controller should be applied on a higher level (node that wraps all form fields)

@pySilver  I think we would keep any wider (non-panel) attrs out of the scope of these changes. Once the dust settles we can look at other parts of the application if there is a need. 

I know that all forms have `data-form` (or similar) on them, so I think it makes sense to review how these are added at some point in the future.

File: wagtail/admin/panels/base.py
Comment (pySilver) [inline wagtail/admin/panels/base.py:86]: for default `attrs=None` there should be 

```
if attrs is None:
    attrs = {}
```

**Comment by pySilver** [comment]
@salty-ivy if that makes your life any easier I've merged your changes into a new branch made of `stable/5.0.x`  here https://github.com/pySilver/wagtail/tree/panel-attributes

**Comment by salty-ivy** [comment]
@pySilver  thanks a lot man, can you DM me on slack we can discuss more about it there this PR already got tons of comment!

and again GREAT thanks for keeping up with this, I would be picking this in no time

**Comment by salty-ivy** [comment]
@lb-  @pySilver   I think most of the things were already done in this only few tests and new `data-form` support is left.

still one major thing I need to look up for is how to support `attrs` in templates without using `self` I believe last time I checked it was working somehow 🫨

**Comment by pySilver** [comment]
I'm playing with this functionality and some stimulus controllers and what I'm missing is attributes on a little bit higher level, specifically at `wagtailadmin/shared/panel.html`. The most common feature is to show/hide sections based on other fields which requires an ability to set attributes at panel level

`<section class="w-panel {{ classname }}" id="{{ panel_id }}" {% if heading %}aria-labelledby="{{ heading_id }}"{% endif %} data-panel>`

I'm wondering how we can handle it without confuse end user? I'm aware we can simply solve it on JS level but Stimulus is all about making less JS code with as general approaches as possible.

What you guys think?

File: wagtail/admin/templates/wagtailadmin/panels/multi_field_panel_child.html
Comment (pySilver) [inline wagtail/admin/templates/wagtailadmin/panels/multi_field_panel_child.html:3]: @salty-ivy there is a bug/typo in the quote position. it should be: 

```<div class="w-panel__wrapper {{ child.classes|join:' ' }}" {% include "wagtailadmin/shared/attrs.html" with attrs=self.attrs %}>```

Comment (salty-ivy) [reply wagtail/admin/templates/wagtailadmin/panels/multi_field_panel_child.html:3]: typo! will fix it.

File: wagtail/admin/templates/wagtailadmin/panels/comments/comment_panel.html
Comment (zerolab) [inline wagtail/admin/templates/wagtailadmin/panels/comments/comment_panel.html:2]: best to pass `attrs` in the context data. using `self` can introduce incompatibilities. see #10538 as an example

Comment (salty-ivy) [reply wagtail/admin/templates/wagtailadmin/panels/comments/comment_panel.html:2]: Yes that wasn't seems to working i dont know for some reason 

@lb- found this way which we are initially using just for the sake of the progress we would definitely need to find out some what was happening

Comment (zerolab) [reply wagtail/admin/templates/wagtailadmin/panels/comments/comment_panel.html:2]: you need to add it via the panel's `BoundPanel.get_context_data()`

**Comment by pySilver** [comment]
@salty-ivy @lb- I have a few more thoughts after working with this PR. 

In my opinion, we should handle MultiFieldPanel attributes differently. Take a look at the code and the result:

```python
content_panels = [
        TitleFieldPanel("title"),
        PageChooserPanel("category", page_type="catalog.CategoryPage"),
        MultiFieldPanel(
            [
                FieldPanel(
                    "term_type",
                    widget=forms.Select(
                        attrs={
                            "data-action": "toggle-select#changed",
                            "data-toggle-select-target": "select",
                        }
                    ),
                ),
                FieldPanel(
                    "content",
                    classname="search_term_content",
                    attrs={
                        "data-values": "landing_page",
                        "hidden": True,
                    },
                ),
                FieldPanel(
                    "description",
                    classname="search_term_description",
                    attrs={
                        "data-values": "search_results",
                        "hidden": True,
                    },
                ),
            ],
            attrs={"data-multitest": "test"},
        ),
    ]
```

CleanShot 2023-06-23 at 14 20 08@2x

I'm surprised it is repeated on every panel instead of being applied to a wrapper once. If I'd need it to be repeated I'd define such `attrs` on a  FieldPanel level.

**Comment by lb-** [comment]
Yeah. I think that's a result of using self.attrs. When I looked at the code (over a month ago now) I think I flagged that we probably would not be able to support some panels that do not render their own wrapping div/container. 

I recall the MultiFieldPanel and a few others just render an array of children without owning the outer div. 

Anyway, I agree that the attrs should not be repeated in this case, thanks for the checks.

**Comment by pySilver** [comment]
@lb- I was thinking about how to overcome this and I've found some solution but it conflicts with default `data-*` (for instance `data-field-wrapper`) that would be repeated in unexpected places. So, either UI code that depends on these few data attributes should be refactored along with these changes or we can leave the default attributes in templates as it is in `main` at the moment but add rendering of custom attributes only. Literally something like that:

```
data-attr-that-is-expected-by-wagtail-here {% include "wagtailadmin/shared/attrs.html" with attrs=self.attrs %}
```

**Comment by pySilver** [comment]
@zerolab @lb-  I'm failing to understand what's wrong with `self.attrs` and how it differs from `self.classes` used in the same template. As you can see in my fork (link below to comparison below) at some places it has to be `child.attrs`. Is that related to `dict` being a mutable structure?

@lb- @salty-ivy I think that after battle testing different setups with my Stimulus controllers, I came up with the sane defaults setup where one can set custom attributes on every level without breaking Wagtail UI code that expects some attributes to be placed at some specific html nodes. That let me retire **BASE_ATTR** property and minimize the number of modified repo files.

Here is the setup that I'm testing against:

```python
content_panels = [
        TitleFieldPanel("title"),
        PageChooserPanel("category", page_type="catalog.CategoryPage", attrs={"data-pagechooser-level": ""}),
        FieldPanel(
            "description",
            attrs={
                "data-top-field-level": True,
            },
        ),
        MultiFieldPanel(
            [
                FieldPanel(
                    "term_type",
                    widget=forms.Select(
                        attrs={
                            "data-widget-level": True,
                        }
                    ),
                ),
                FieldPanel(
                    "content",
                    attrs={
                        "data-top-field-level": True,
                    },
                ),
            ],
            attrs={"data-multifield-level": "test"},
        ),
    ]
```

Output:
CleanShot 2023-06-23 at 19 25 00

Diff agains `stable/5.0.x` is here: https://github.com/wagtail/wagtail/compare/stable/5.0.x...pySilver:wagtail:panel-attributes?expand=1

Branch: https://github.com/pySilver/wagtail/tree/panel-attributes

**Comment by lb-** [comment]
The goal of the BASE_ATTRS on classes was two fold, allow the existing (if used) attrs that already come with panels to move to the classes, not the templates. 

This in turn allows these to be overridden, reused, extended as needed. By default, they would just be appended to, but the Dev has more power to change things if they want. 

For example, without the base attrs approach there is no way to override the defaults like data-panel. 

This becomes even more important when you consider the data-controller/data-action attributes. Stimulus intentionally allows these to be used with multiple controllers in mind. 

Example usage. You have a panel that is collapsible by default, using data-controller="w-expand" (or Similar in the future). Through your customisations you want to add another controller to make it data-controller="w-panel my-custom-controller". Without the base attrs there would be no way to override the data-controller attribute set in the template, we could leverage class inheritance and calling super but I felt that it could be a common enough use case we could put these as a property on the class itself. 

Hope that context helps a bit. I'm thinking the easy customisations should be easy and complex ones should be possible (without having to always reach for a complete template override).

**Comment by pySilver** [comment]
@lb- I totally agree with the idea. But there is one problematic attribute that ruins it and it is `data-field-wrapper`. 

Actually, I took a look at how this `data-field-wrapper` is used in sources and it looks like it doesn't matter if it is specified on `.w-field__wrapper` or at the `.w-panel__wrapper`. It looks like it's safe to change its default rendering position or even rename it.

I'm also wondering if what we are doing here should be reflected in:

```
wagtail/client/src/components/StreamField/blocks/FieldBlock.js
wagtail/client/src/entrypoints/contrib/table_block/table.js
```

since these data attributes are mentioned in these files. For instance:

```js
const dom = $(`
      <div class="w-field__wrapper" data-field-wrapper>
        <div class="${h(this.blockDef.meta.classname)}" data-field>
          <div class="w-field__errors" id="${prefix}-errors" data-field-errors>
            <svg class="icon icon-warning w-field__errors-icon" aria-hidden="true" hidden><use href="#icon-warning"></use></svg>
          </div>
          <div class="w-field__help" id="${prefix}-helptext" data-field-help></div>
          <div class="w-field__input" data-field-input>
            <div data-streamfield-widget></div>
          </div>
        </div>
      </div>
    `);
```

**Comment by pySilver** [comment]
Beside that I think data attributes that come from class definition (data-panel, data-tabs) at least at the moment are very different than what developer might want to define. Like I've said - I've tried to use the original code from this PR and this makes Stimulus controllers useless since we are back to writing DOM CSS queries in order to find specific  node, instead of simply using targets:

The most basic and very common example is the dynamic forms where sections are shown/hidden based on some field state. That is simply impossible without case-specific JS in original PR because we cannot set data attributes on specific level that we want. And if we make it possible then, our attributes from BASE_ATTR magically appears not where we expect them.

**Comment by pySilver** [comment]
@lb- @salty-ivy  what if we keep base attrs feature but without any defaults at the moment? Existing data attributes aren't related to Stimulus controllers and may stay in templates until they migrated to Stimulus. This would solve all issues:

* opens doors for custom panels with default attributes that can be overridden 
* keeps existing code working 
* gives flexible control on putting data attrs on any desired level

What you think?

**Comment by pySilver** [comment]
@lb- @salty-ivy I've brought back BASE_ATTRS into my fork.

@salty-ivy I don't want to take over your work :) It is just I urgently need this feature for development, so I'm kinda working on every possible approach to get it )

**Comment by salty-ivy** [comment]
No issues see if you can pull out anything here.
Anyways I will be joining from from Monday !!

**Comment by pySilver** [comment]
@salty-ivy I guess you are still busy?

**Comment by salty-ivy** [comment]
yup here, @pySilver  really sorry for late reply I was busy with other PRs ( on Willow ) didn't receive your notification.

**Comment by pySilver** [comment]
@salty-ivy I've just merged the changes against the current wagtail:main here https://github.com/pySilver/wagtail/tree/panel-attrs I hope it will help you finish your PR :)

**Comment by lb-** [comment]
No problems at all @salty-ivy - please do not worry - if it happens it happens.

**Comment by pySilver** [comment]
@lb yeah. I'm happy to finish it. I guess I should open a separate PR with code synced into current main, right?

**Comment by lb-** [comment]
> @lb yeah. I'm happy to finish it. I guess I should open a separate PR with code synced into current main, right?

Yep, that will be best.

**Comment by pySilver** [comment]
@lb- @salty-ivy here it is #10629

**Comment by zerolab** [comment]
Thank you @pySilver. Closing in favour of #10629

**Pull request closed without merging**
